airflow/providers/google/cloud/operators/dataproc.py

```diff
@@ -602,22 +609,50 @@ def execute(self, context: Context) -> dict:
                 raise
             self.log.info("Cluster already exists.")
             cluster = self._get_cluster(hook)
-
-        # Check if cluster is not in ERROR state
-        self._handle_error_state(hook, cluster)
-        if cluster.status.state == cluster.status.State.CREATING:
-            # Wait for cluster to be created
-            cluster = self._wait_for_cluster_in_creating_state(hook)
-            self._handle_error_state(hook, cluster)
-        elif cluster.status.state == cluster.status.State.DELETING:
-            # Wait for cluster to be deleted
-            self._wait_for_cluster_in_deleting_state(hook)
-            # Create new cluster
-            cluster = self._create_cluster(hook)
+        if not self.deferrable:
+            # Check if cluster is not in ERROR state
             self._handle_error_state(hook, cluster)
-
+            if cluster.status.state == cluster.status.State.CREATING:
+                # Wait for cluster to be created
+                cluster = self._wait_for_cluster_in_creating_state(hook)
+                self._handle_error_state(hook, cluster)
+            elif cluster.status.state == cluster.status.State.DELETING:
+                # Wait for cluster to be deleted
+                self._wait_for_cluster_in_deleting_state(hook)
+                # Create new cluster
+                cluster = self._create_cluster(hook)
+                self._handle_error_state(hook, cluster)
+        else:
+            self.defer(
+                trigger=DataprocClusterTrigger(
+                    cluster_name=self.cluster_name,
+                    project_id=self.project_id,
+                    region=self.region,
+                    gcp_conn_id=self.gcp_conn_id,
+                    impersonation_chain=self.impersonation_chain,
+                    poll_interval=self.poll_interval,
+                ),
+                method_name="execute_complete",
+            )
         return Cluster.to_dict(cluster)
 
+    def execute_complete(self, context, event=None) -> None:
+        """
+        Callback for when the trigger fires - returns immediately.
+        Relies on trigger to throw an exception, otherwise it assumes execution was
+        successful.
+        """
+        if event is None:
+            raise AirflowException("Cluster failed.")
+        cluster_state = event["cluster_state"]
+        cluster_name = event["cluster_name"]
+
+        if cluster_state == ClusterStatus.State.ERROR:
+            raise AirflowException("Cluster is in ERROR state.")
+        if cluster_state == ClusterStatus.State.DELETING:
+            raise AirflowException(f"Cluster is being deleted:\n{cluster_name}")
+        self.log.info("%s completed successfully.", self.task_id)
```

Comment:
you should push the cluster details in xcom as sync version of operator does

Reply:
fix